tonic/src/service/interceptor.rs

```diff
@@ -52,6 +54,97 @@ where
     }
 }
 
+/// Async version of `Interceptor`.
+pub trait AsyncInterceptor {
```

Comment:
This feels almost exactly like the tower service trait, what is the advantage and why don't we just use the tower traits?

Reply:
Hi @LucioFranco , sorry for letting this PR sit for so long. I've merged the latest master into my branch, updated the code to work with those changes, and refactored a bit to reduce code duplication.

Note that I had to constrain ReqBody to implement Default. Seems like that should be ok (and ResBody already had that constraint).

In response to your question...

I would love to use something in Tower rather than create all of this gnarly generics+futures code, but I couldn't find anything that will do the trick. You could ask the same thing about the non-async interceptor, too (written by @davidpdrsn   ?). I think the main reason is that Tower doesn't have anything that both (a) modifies a request and (b) allows an error response to be returned immediately.

Tower has [`service_fn`](https://tower-rs.github.io/tower/tower/fn.service_fn.html) which makes an async function into a Tower Service, but it doesn't wrap an inner service, so I don't think it can be used as a middleware layer.

Looking at the [`tower::util::ServiceExt`](https://docs.rs/tower/latest/tower/util/trait.ServiceExt.html) trait, there are are some promising methods that don't quite fit the async_interceptor case:

 - [`map_request`](https://docs.rs/tower/latest/tower/util/trait.ServiceExt.html#method.map_request) transforms the request and wraps the service, but it doesn't accept async functions, and it cannot skip calling the wrapped service.
 - [`filter_async`](https://docs.rs/tower/latest/tower/util/trait.ServiceExt.html#method.filter_async) is doesn't transform the request, but is otherwise close to what we want.

I also spent quite a bit of time trying to simplify this code with async functions, but since their return types are generated at compile time, it's hard to make them play nicely with the `Service` trait since it requires explicitly specifying the `Future` associated type, and we can't set `Future = impl Future<Output = ...>` until [this rust RFC](https://github.com/rust-lang/rust/issues/63063) is accepted and shipped.

Reply:
The main advantage is that this uses tonic's `Request` and `Status` types. Whereas tower middleware have to use the types from `http`.